File: ch4/python/code4_7.py
def waysToScore(n):
  # 파이썬 리스트는 아래와 같이 일괄 초기화가 가능합니다.
  arr = [0] * (n + 1)

  arr[0] = 1  # 0점까지의 경우의 수는 아무것도 하지 않는 1가지입니다.

  for i in range(1, n + 1):
    if i - 3 >= 0:
      arr[i] += arr[i - 3]
    if i - 5 >= 0:
      arr[i] += arr[i - 5]
    if i - 10 >= 0:
      arr[i] += arr[i - 10]
  return arr[n]    

N = int(input('목적 점수를 입력하세요 : '))
print('%d점 까지의 경우의 수는 %d 입니다.' % (N, waysToScore(N)))

File: ch4/c/code4_7.c
#include <stdio.h>

int waysToScore(int n)
{
  // 변수를 사용해 선언한 배열은 초기화를 할 수 없습니다. 
  // 즉 int arr[n + 1] = {0}은 컴파일 에러가 발생합니다.
  int arr[n + 1];	
  for(int i = 0; i <= n; i++)
    arr[i] = 0;

  arr[0] = 1;		// 0점까지의 경우의 수는 아무것도 하지 않는 1가지입니다.

  for(int i = 1; i <= n; i++)
  {
    if(i - 3 >= 0)
      arr[i] += arr[i - 3];
    if(i - 5 >= 0)
      arr[i] += arr[i - 5];
    if(i - 10 >= 0)
      arr[i] += arr[i - 10];
  }
  return arr[n];
}

int main()
{
  printf("목적 점수를 입력하세요 : ");
  int N;
  scanf("%d", &N);
  printf("%d점까지의 경우의 수는 %d입니다.\n", N, waysToScore(N));
  return 0;
}